src/lib.rs

```diff
@@ -2727,53 +2722,71 @@ impl<
     /// Get the status of a transaction.
     pub async fn transaction_status(
         &self,
-        receipt: &TransactionReceipt<L>,
+        uid: &TransactionUID<L>,
     ) -> Result<TransactionStatus, KeystoreError<L>> {
-        let KeystoreSharedState { state, model, .. } = &mut *self.mutex.lock().await;
-        state.transaction_status(model, receipt).await
+        let KeystoreSharedState {
+            state: _, model, ..
+        } = &mut *self.mutex.lock().await;
+        Ok(model.transactions.get(uid)?.status())
+    }
+
+    /// A future which completes when a transaction we didn't submit completes
+    pub async fn await_foreign_transaction(
+        &self,
+        receipt: TransactionReceipt<L>,
+    ) -> Result<TransactionStatus, KeystoreError<L>> {
+        let mut guard = self.mutex.lock().await;
+        let KeystoreSharedState {
+            state: _,
+            model: _,
+            txn_subscribers: _,
+            pending_foreign_txns,
+            ..
+        } = &mut *guard;
+
+        let status = self.transaction_status(&receipt.uid).await?;
+        if status.is_final() {
+            Ok(status)
+        } else {
+            let (sender, receiver) = oneshot::channel();
+            // Transaction uids are unique only to a given keystore, so if we're trying to track
+            // somebody else's transaction, the best we can do is wait for one of its nullifiers
+            // to be published on the ledger.
+            pending_foreign_txns
+                .entry(receipt.fee_nullifier)
+                .or_insert_with(Vec::new)
+                .push(sender);
+            drop(guard);
+            receiver.await.map_err(|_| KeystoreError::<L>::Cancelled {})
+        }
```

Comment:
I thought we were getting rid of this behavior. If we do, we can also get rid of `TransactionReceipt` (because we would only ever use `receipt.uid`) and some of the complicated logic with `pending_foreign_txns` in the event handling thread. Any reason to keep this?

Reply:
Only reason to keep receipt was because the CLI uses receipt in it's output but I guess if we have the hash we can do whatever we need.

This function we can get rid of.  I don't really know why I kept it

Reply:
Nice! This is one of the parts I always hated most. And yeah, the CLI can just output the transaction hash (which makes us a little bit more consistent with how most Ethereum stuff works, anyway)